feature: add `#[salsa::interned_sans_lifetime]` macro

davidbarsky opened the pull request:
(This consolidates #618, #620, and #623 into a single PR, logical PR.)

Here's what this solves:
1. the `'db` lifetime on interned structs makes migrating some rust-analyzer's existing interned values *really* challenging.
2. rust-analyzer wants to fiddle with the underlying ID.
3. We'd like to be able to fetch the associated, interned data without having the wrapped, interned struct on hand.

We'll probably port rust-analyzer over the new, idiomatic API instead, but this makes the migration a lot easier!

MichaReiser commented:
We discussed that we want to discourage this pattern but it would also be nice if we can avoid duplicating the interned macro. Would it be possible to gate the functionality behind a feature? E.g. that the `interned` macro only supports a `no_lifetime` argument if `salsa_macros` is used with the `compat` feature.

Veykril commented:
Given the goal is for this to only ease the transition for r-a it should be fine (and honestly preferred in my eyes) to gate this behind a feature until r-a no longer needs it (at which point we can remove it again).

MichaReiser commented:
It's hard to tell for me what the difference is between the two macros but would it be possible to change `$db_lt` to either be `<'db'>` (so that it includes the `<..>` ` or an empty string if `no_lifetime` is used.

davidbarsky commented:
> It's hard to tell for me what the difference is between the two macros

I do too, but difftastic helped here. I don't think it's a particularly useful paste, but if you have it handy, `difft components/salsa-macro-rules/src/setup_interned_struct.rs components/salsa-macro-rules/src/setup_interned_struct_sans_lifetime.rs` will show the difference. The primary differences are:
1. The addition of a `StructData` parameter to the macro. I didn't want pull in `paste`, so I just made the ident in the proc macro.
2. The addition of the `$Id` parameter, to allow the ID overriding behavior we need in rust-analyzer.
3. Moving `salsa::plumbing::interned::Configuration` out of the `const _: () = {...} ` block in order to get the data through the underlying ID alone.
4. The removal of lifetimes and a few select additions of `'static`, as needed.

I honestly recommend cloning this PR and diffing via difftastic.

> would it be possible to change `$db_lt` to either be `<'db>` (so that it includes the `<..>`) or an empty string if `no_lifetime` is used? 

Not that I'm aware of, at least with `macro_rules!`-based macros. I don't think we can invoke `maybe_*`-style macros in the type position. From a "how can we unify these macros", I _think_ it'd require moving things away from `salsa-macro-rules` into a (mostly) pure `quote!`-based approach.

davidbarsky commented:
I fixed the rustdoc failures, but I can't reproduce the test failure w.r.t to cancellation on my Mac. That's a bit concerning, but I think I'll try and debug this tomorrow.

Veykril commented:
I question the need for the `id =` part actually, I'll raise the why in the rust-analyzer zulip but just so people know. I don't think r-a needs that part.

davidbarsky commented:
> I fixed the rustdoc failures, but I can't reproduce the test failure w.r.t to cancellation on my Mac. That's a bit concerning, but I think I'll try and debug this tomorrow.

Lukas found and fixed the issue in https://github.com/salsa-rs/salsa/pull/637/.

davidbarsky commented:
Closing in favor of https://github.com/salsa-rs/salsa/pull/661, which has substantially less duplication and removes support for [overriding IDs](https://github.com/salsa-rs/salsa/pull/632#issuecomment-2558538153).
